lightning/src/util/enforcing_trait_impls.rs

```diff
@@ -72,12 +92,22 @@ impl ChannelKeys for EnforcingChannelKeys {
 		Ok(self.inner.sign_counterparty_commitment(commitment_tx, secp_ctx).unwrap())
 	}
 
-	fn sign_holder_commitment<T: secp256k1::Signing + secp256k1::Verification>(&self, commitment_tx: &HolderCommitmentTransaction, secp_ctx: &Secp256k1<T>) -> Result<Signature, ()> {
-		self.verify_holder_commitment_tx(commitment_tx, secp_ctx);
-
-		// TODO: enforce the ChannelKeys contract - error if this commitment was already revoked
-		// TODO: need the commitment number
-		Ok(self.inner.sign_holder_commitment(commitment_tx, secp_ctx).unwrap())
+	fn sign_holder_commitment<T: secp256k1::Signing + secp256k1::Verification>(&self, holder_commitment_tx: &HolderCommitmentTransaction, secp_ctx: &Secp256k1<T>) -> Result<Signature, ()> {
+		let trusted_tx = self.verify_holder_commitment_tx(holder_commitment_tx, secp_ctx);
+
+		let revoked = self.revoked_commitment.lock().unwrap();
+		let commitment_number = trusted_tx.commitment_number();
+		if *revoked - 1 != commitment_number && *revoked - 2 != commitment_number {
+			let msg = format!("can only sign the next two unrevoked commitment numbers, revoked={} vs requested={} for {}",
+			       *revoked, commitment_number, self.inner.commitment_seed[0]);
+			if self.err_on_sign_revoked_holder_tx {
+				println!("{}", msg);
```

Comment:
Was this just for testing? This is just for testing for now, so kinda whatever, but if we ever make this struct pub all printing has to go via a logger, so maybe leave a comment here?

Reply:
Well, this struct *is* pub.

But since the flag is mostly for testing for now, there's no strong reason for the print, so removing it.

Reply:
All of `util::enforcing_trait_impls` is pub only for test,fuzz,_test_utils.

Reply:
got it